nuclio/export.py

```diff
@@ -95,34 +101,62 @@ def from_notebook_node(self, nb, resources=None, **kw):
             config['metadata']['name'] = normalize_name(name)
         config['spec']['handler'] = handler_name()
-
-        io = StringIO()
-        print(header(), file=io)
+        function_name = environ.get(env_keys.function_name)
+        if not function_name:
+            # function name must differ from empty name
+            function_name = None
+        seen_function_name = ""
+        function_buffers = {
+            function_name: {
+                'closed': False,
+                'codes': [],
+            },
+            "": {
+                'closed': False,
+                'codes': [],
+            },
+        }
 
         for cell in filter(is_code_cell, nb['cells']):
             code = cell['source']
-            if has_ignore(code):
+            match = has_ignore(code)
+            if match:
+                curr_name = match.group('name')
+                if curr_name == "" \
+                        and not function_buffers[function_name]['closed']:
+                    function_buffers[function_name]['codes'].append(code)
+                elif curr_name == function_name:
+                    function_buffers[""]['closed'] = True
+                    seen_function_name = curr_name
                 continue
 
-            if has_end(code):
-                break
-
-            if has_start(code):
-                # if we see indication of start, we ignore all previous cells
-                io = StringIO()
-                print(header(), file=io)
-
-            code = filter_comments(code)
-            if not code.strip():
+            match = has_end(code)
+            if match:
+                curr_name = match.group('name')
+                if curr_name in [function_name, ""]:
+                    function_buffers[curr_name]['closed'] = True
+                    if curr_name == function_name:
+                        # if we see end with function name we can break
+                        seen_function_name = curr_name
+                        break
                 continue
 
-            lines = code.splitlines()
-            if cell_magic in code:
-                self.handle_cell_magic(lines, io, config)
-                continue
-
-            self.handle_code_cell(lines, io, config)
-
+            match = has_start(code)
+            if match:
+                # if we see indication of start, we ignore all previous cells
+                curr_name = match.group('name')
+                if curr_name in [function_name, ""]:
+                    function_buffers[curr_name]['codes'] = []
+                    if curr_name == function_name:
+                        function_buffers[""]['closed'] = True
+                        seen_function_name = curr_name
+
+            for function_buffer in function_buffers.values():
+                if not function_buffer['closed']:
+                    function_buffer['codes'].append(code)
+
+        io = self.write_codes(function_buffers[seen_function_name]['codes'],
```

Comment:
I only now understand that there can be several weird edge cases here, like:
```
# mlrun: start-code my-function
<some code>
# mlrun: start-code my-function
<some other code>
# mlrun: end-code my-function
```
or
```
# mlrun: start-code my-function
<some code>
# mlrun: end-code my-function
<some other code>
# mlrun: end-code my-function
```
It's a bit hard to imagine all of them in the head while reviewing the code, so I suggest that you'll list all of the possible cases, verify you know the expected behavior (and if not let's talk about it me you and Oded), and add tests for all of them (you can start from the tests and go full TDD)